python/ray/experimental/client/client_pickler.py

```diff
@@ -0,0 +1,100 @@
+import cloudpickle
+import io
+import sys
+
+from typing import NamedTuple
+
+from ray.experimental.client.common import ClientObjectRef
+from ray.experimental.client.common import ClientActorHandle
+from ray.experimental.client.common import ClientActorRef
+from ray.experimental.client.common import ClientRemoteFunc
+import ray.core.generated.ray_client_pb2 as ray_client_pb2
+
+if sys.version_info < (3, 8):
+    try:
+        import pickle5 as pickle  # noqa: F401
+    except ImportError:
+        import pickle  # noqa: F401
+else:
+    import pickle  # noqa: F401
+
+PickleStub = NamedTuple("PickleStub", [("type", str), ("client_id", str),
+                                       ("ref_id", bytes)])
+
+
+class ClientPickler(cloudpickle.CloudPickler):
+    def __init__(self, client_id, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.client_id = client_id
+
+    def persistent_id(self, obj):
+        if isinstance(obj, ClientObjectRef):
+            return PickleStub(
```

Comment:
Can't we make this a class function in ClientObjectRef, ClientActorHandle, ClientRemoteFunc that receives the client id and object and returns the `PickleStub`?

Reply:
Maybe all these classes should inherit a common class so that adding a new Client instance won't require changing anything here.

Reply:
I thought about the first notion -- I'd like to leave it as is, because then the protocol is a little more clear. We can't add the same ones to the server pickler (without changing the ray objects) so this way the two files should mirror each other, instead of one having implementations in the objects and the other following a switch.

Inheriting a common class wouldn't help. There's no 'baseclass' PickleStub -- they should be explicitly implemented.

Which goes back to your first comment -- making the protocol implementation live on the object instead of in the file, but I'm opting for having the protocol in one file.